File: NDNHealthCloud/interestpacket.js
/** 
 * File contains code for creating 
 * NDN-compliant interest packets
 **/

var NameField = require('./namefield');
var StringConst = require('./string_const').StringConst;
var Utils = require('./utils').Utils;

exports.InterestPacket = function () {
   
   return { 

        LIFETIME_CONST : 100,

        InterestPacket: function (userDataID, sensorID, timestring, processID, ipAddr) {

            console.log("TOIMESTRING: " + timestring);
            console.log("string const current time: " + StringConst.CURRENT_TIME);

             // if current time requested, provide it
            if (timestring === StringConst.CURRENT_TIME) {
                console.log("if");
                timestring = Utils.getCurrentTime();
            } else {
                console.log("else");
            }

            this.nameField = NameField.NameField();
            this.nameField.NameField(userDataID, sensorID, timestring, processID, ipAddr);
            this.NON_NEG_INT_CONST = 0; // TODO - rework
        },

        /**
        Nonce ::=
        --------------
        NONCE-TYPE TLV-LENGTH(=4) BYTE{4}
        --------------

         "The Nonce carries a randomly-generated 4-octet long byte-string.
         The combination of Name and Nonce should uniquely identify an Interest packet.
         This is used to detect looping Interests."
        **/
        createNonce: function () {

            var max = 255; // 2^8 -1 (max 8 bit number)
            var min = 0; // min unsigned 8 bit number
            var randomInt = Math.ceil(Math.random() * (max-min) + min);

            var content = randomInt.toString(); // first octet

            randomInt = Math.ceil(Math.random() * (max-min) + min);
            content += randomInt.toString(); // second octet

            randomInt = Math.ceil(Math.random() * (max-min) + min);
            content += randomInt.toString(); // third octet

            randomInt = Math.ceil(Math.random() * (max-min) + min);
            content += randomInt.toString(); // fourth octet

            return "NONCE-TYPE " + (content.length).toString() + " " + content;    
        },
        
        /**
        Selectors ::=
        --------------
         SELECTORS-TYPE TLV-LENGTH
         MinSuffixComponents?
         MaxSuffixComponents?
         PublisherPublicKeyLocator?
         Exclude?
         ChildSelector?
         MustBeFresh?
        --------------
        **/
        createSelectors: function () {
            var content = this.createMinSuffixComponents();
            content += " " + this.createMaxSuffixComponents();
            content += " " + this.createPublisherPublicKeyLocator();
            content += " " + this.createExclude();
            content += " " + this.createChildSelector();
            content += " " + this.createMustBeFresh();

            return "SELECTORS-TYPE " + (content.length).toString() + " " + content;
        },

        /**
         MinSuffixComponents ::=
         --------------
         MIN-SUFFIX-COMPONENTS-TYPE TLV-LENGTH
         nonNegativeInteger
         --------------
         */
        createMinSuffixComponents: function () {
            var content = "0"; // TODO - rework later
            return "MIN-SUFFIX-COMPONENTS-TYPE " + (content.length).toString() + " " + content;
        },

        /**
         MaxSuffixComponents ::=
         --------------
         MAX-SUFFIX-COMPONENTS-TYPE TLV-LENGTH
         nonNegativeInteger
         --------------
         */
        createMaxSuffixComponents: function () {
            var content = "0"; // TODO - rework later
            return "MAX-SUFFIX-COMPONENTS-TYPE " + (content.length).toString() + " " + content;
        },

        /**
         PublisherPublicKeyLocator ::=
         --------------
         KeyLocator
         --------------
         */
        createPublisherPublicKeyLocator: function () {
            return "null"; // TODO - rework later
        },

        /**
         Any ::=
         --------------
         ANY-TYPE TLV-LENGTH(=0)
         --------------
         */
        createAny: function () {
            return "ANY-TYPE 0"; // TODO - rework
        },

        /**
         Exclude ::=
         --------------
         EXCLUDE-TYPE TLV-LENGTH Any? (NameComponent (Any)?)+
         --------------
         */
        createExclude: function () {
            var content = this.createAny();
            return "EXCLUDE-TYPE 0 " + (content.length).toString() + " " + content; // TODO - rework later
        },

        /**
         ChildSelector ::=
         --------------
         CHILD-SELECTOR-TYPE TLV-LENGTH
         nonNegativeInteger
         --------------
         */
        createChildSelector: function () {
            var content = "0"; // TODO - rework
            return "CHILD-SELECTOR-TYPE " + (content.length).toString() + " " + content;
        },

        /**
         MustBeFresh ::=
         --------------
         MUST-BE-FRESH-TYPE TLV-LENGTH(=0)
         --------------
         */
        createMustBeFresh: function () {
            return "MUST-BE-FRESH-TYPE 0"; // TODO - rework
        },

        /**
        InterestLifeTime ::=
        --------------
        INTEREST-LIFETIME-TYPE TLV-LENGTH nonNegativeInteger
        --------------
        **/
        createInterestLifetime: function () {
            var content = this.LIFETIME_CONST.toString(); // TODO - add user-selected interval
            return "INTEREST-LIFETIME-TYPE " + (content.length).toString() + " " + content;
        },

        /**
        Scope ::=
        --------------
        SCOPE-TYPE TLV-LENGTH nonNegativeInteger
        --------------
        */
        createScope: function () {
            var content = "0"; // TODO - rework
            return "SCOPE-TYPE " + (content.length).toString() + " " + content;
        },

        /**
         INTEREST ::=
         --------------
         INTEREST-TYPE TLV-LENGTH
         Name
         Selectors?
         Nonce
         Scope?
         Interest Lifetime?
         --------------
         **/
        createINTEREST: function () {
            var content = this.nameField.createName();
            content += " " + this.createSelectors();
            content += " " +  this.createNonce();
            content += " " + this.createScope();
            content += " " + this.createInterestLifetime();

            return "INTEREST-TYPE " + (content.length).toString() + " " + content;
        },

        toString: function () {
            return this.createINTEREST();
        }
    }
};

File: PHINetDroid/app/src/main/java/com/ndnhealthnet/androidudpclient/Packet/InterestPacket.java
package com.ndnhealthnet.androidudpclient.Packet;

import com.ndnhealthnet.androidudpclient.Utility.StringConst;
import com.ndnhealthnet.androidudpclient.Utility.Utils;

import java.util.Random;

/**
 * Class creates an NDN-Compliant Interest packet in the form of a string.
 *
 * Each component of a packet has its own creation method.
 *
 * Specification found here: http://named-data.net/doc/ndn-tlv/data.html
 */
public class InterestPacket {

    // TODO - only send certain portions of InterestPacket (be able to avoid optional)
        // TODO - do this through constructor, etc

    private NameField nameField;
    final private int LIFETIME_CONST = 100;

    /**
     * @param userDataID specifies data producer
     * @param sensorID specifies health-sensor type (e.g., heart sensor)
     * @param timeString specifies when packet was created
     * @param processID specifies what process should be invoked upon reception (e.g., store in cache)
     */
    public InterestPacket(String userDataID, String sensorID, String timeString, String processID) {

        // if current time requested, provide it
        if (timeString.equals(StringConst.CURRENT_TIME)) {
            timeString = Utils.getCurrentTime();
        }

        nameField = new NameField(userDataID, sensorID, timeString, processID);
    }

    /**
     * Nonce ::=
     * --------------
     * NONCE-TYPE TLV-LENGTH(=4) BYTE{4}
     * --------------
     *
     * "The Nonce carries a randomly-generated 4-octet long byte-string.
     * The combination of Name and Nonce should uniquely identify an Interest packet.
     * This is used to detect looping Interests."
     *
     * @return Nonce as definition above shows (see NDN specification)
    **/
    String createNonce() {

        Random r = new Random();

        String content = Integer.toString(r.nextInt(255)); // first octet
        content += Integer.toString(r.nextInt(255)); // second octet
        content += Integer.toString(r.nextInt(255)); // third octet
        content += Integer.toString(r.nextInt(255)); // fourth octet

        return "NONCE-TYPE " + Integer.toString(content.length()) + " " + content;    
    }
    
    /**
     * Selectors ::=
     * --------------
     * SELECTORS-TYPE TLV-LENGTH
     * MinSuffixComponents?
     * MaxSuffixComponents?
     * PublisherPublicKeyLocator?
     * Exclude?
     * ChildSelector?
     * MustBeFresh?
     * --------------
     *
     * @return Selectors as definition above shows (see NDN specification)
    **/
    String createSelectors() {
        String content = createMinSuffixComponents();
        content += " " + createMaxSuffixComponents();
        content += " " + createPublisherPublicKeyLocator();
        content += " " + createExclude();
        content += " " + createChildSelector();
        content += " " + createMustBeFresh();

        return "SELECTORS-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * MinSuffixComponents ::=
     * --------------
     * MIN-SUFFIX-COMPONENTS-TYPE TLV-LENGTH
     * nonNegativeInteger
     * --------------
     *
     * @return MinSuffixeComponents as definition above shows (see NDN specification)
     */
    String createMinSuffixComponents() {
        String content = "0"; // TODO - rework
        return "MIN-SUFFIX-COMPONENTS-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * MaxSuffixComponents ::=
     * --------------
     * MAX-SUFFIX-COMPONENTS-TYPE TLV-LENGTH
     * nonNegativeInteger
     * --------------
     *
     * @return MaxSuffixeComponents as definition above shows (see NDN specification)
     */
    String createMaxSuffixComponents() {
        String content = "0"; // TODO - rework
        return "MAX-SUFFIX-COMPONENTS-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * PublisherPublicKeyLocator ::=
     * --------------
     * KeyLocator
     * --------------
     *
     * @return PublisherPublicKeyLocator as definition above shows (see NDN specification)
     */
    String createPublisherPublicKeyLocator() {
        return "null"; // TODO - rework later
    }

    /**
     * Any ::=
     * --------------
     * ANY-TYPE TLV-LENGTH(=0)
     * --------------
     *
     * @return Any as definition above shows (see NDN specification)
     */
    String createAny() {
        return "ANY-TYPE 0";
    }

    /**
     * Exclude ::=
     * --------------
     * EXCLUDE-TYPE TLV-LENGTH Any? (NameComponent (Any)?)+
     * --------------
     *
     * @return Exclude as definition above shows (see NDN specification)
     */
    String createExclude() {
        String content = createAny(); // TODO - rework
        return "EXCLUDE-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * ChildSelector ::=
     * --------------
     * CHILD-SELECTOR-TYPE TLV-LENGTH
     * nonNegativeInteger
     * --------------
     *
     * @return ChildSelector as definition above shows (see NDN specification)
     */
    String createChildSelector() {
        String content = "0"; // TODO - rework
        return "CHILD-SELECTOR-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * MustBeFresh ::=
     * --------------
     * MUST-BE-FRESH-TYPE TLV-LENGTH(=0)
     * --------------
     *
     * @return MustBeFresh as definition above shows (see NDN specification)
     */
    String createMustBeFresh() {
        return "MUST-BE-FRESH-TYPE 0";
    }

    /**
     * InterestLifeTime ::=
     * --------------
     * INTEREST-LIFETIME-TYPE TLV-LENGTH nonNegativeInteger
     * --------------
     *
     * @return InterestLifeTime as definition above shows (see NDN specification)
    **/
    String createInterestLifetime() {
        String content = Integer.toString(LIFETIME_CONST); // TODO - add user-selected interval
        return "INTEREST-LIFETIME-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * Scope ::=
     * --------------
     * SCOPE-TYPE TLV-LENGTH nonNegativeInteger
     * --------------
     *
     * @return Scope as definition above shows (see NDN specification)
    */
    String createScope() {
        String content = "0"; // TODO - rework
        return "SCOPE-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    /**
     * INTEREST ::=
     * --------------
     * INTEREST-TYPE TLV-LENGTH
     * Name
     * Selectors?
     * Nonce
     * Scope?
     * Interest Lifetime?
     * --------------
     *
     * @return Interest packet as definition above shows (see NDN specification)
     **/
    String createINTEREST() {
        String content = nameField.createName();
        content += " " + createSelectors();
        content += " " +  createNonce();
        content += " " + createScope();
        content += " " + createInterestLifetime();

        return "INTEREST-TYPE " + Integer.toString(content.length()) + " " + content;
    }

    @Override
    public String toString() {
        return createINTEREST();
    }
}
